### PR TITLE
fix flaky integration test about multi profiles

### LINKED ISSUE #125653: Failure cluster [9afae275...]
##### Error text:
```
Failed;Failed;
=== RUN   TestFrameworkHandler_IterateOverWaitingPods/pods_with_same_profile_are_waiting_on_permit_stage
W0612 12:30:02.495091   63674 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0612 12:30:02.495137   63674 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
    framework.go:381: I0612 12:30:02.495421] the scheduler starts to work with those plugins Plugins={"PreEnqueue":{"Enabled":null,"Disabled":null},"QueueSort":{"Enabled":[{"Name":"PrioritySort","Weight":0}],"Disabled":null},"PreFilter":{"Enabled":null,"Disabled":null},"Filter":{"Enabled":null,"Disabled":null},"PostFilter":{"Enabled":null,"Disabled":null},"PreScore":{"Enabled":null,"Disabled":null},"Score":{"Enabled":null,"Disabled":null},"Reserve":{"Enabled":null,"Disabled":null},"Permit":{"Enabled":[{"Name":"fakePermit","Weight":0}],"Disabled":null},"PreBind":{"Enabled":null,"Disabled":null},"Bind":{"Enabled":[{"Name":"DefaultBinder","Weight":0}],"Disabled":null},"PostBind":{"Enabled":null,"Disabled":null},"MultiPoint":{"Enabled":null,"Disabled":null}}
W0612 12:30:02.495645   63674 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0612 12:30:02.495860   63674 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0612 12:30:02.495955   63674 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0612 12:3
```
#### Recent failures:
[6/12/2024, 2:07:33 PM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1800862311091539968)

/kind failing-test
/kind flake

First seen in an unrelated PR, but also occurs in periodic jobs.

/sig scheduling

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind cleanup
/kind flake
/sig scheduling

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/125653
Xref: https://github.com/kubernetes/kubernetes/pull/123686

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -994,14 +995,16 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
...
 			eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme, fakePermit)
 
 			outOfTreeRegistry := frameworkruntime.Registry{
 				fakePermit: newFakePermitPlugin(eventRecorder),
 			}
 
 			_, ctx := ktesting.NewTestContext(t)
-			ctx, cancel := context.WithCancel(ctx)
+			// timeout equals to the permit plugin waiting time.
+			ctx, cancel := context.WithTimeout(ctx, 100*time.Second)
```
Comment (pohly) [inline pkg/scheduler/scheduler_test.go:1007]: Where is that time specified? What if it changes?

File: pkg/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -1049,17 +1052,23 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
...
+					})
+					// Validate the name of pods in waitingPods matches expectations.
+					if actualPodNamesInWaitingPods.Len() != len(tc.expectPodNamesInWaitingPods) ||
+						!actualPodNamesInWaitingPods.HasAll(tc.expectPodNamesInWaitingPods...) {
+						return false, fmt.Errorf("Unexpected waitingPods in scheduler profile %s, expect: %#v, got: %#v", fwk.ProfileName(), tc.expectPodNamesInWaitingPods, actualPodNamesInWaitingPods.List())
+					}
 				}
+				return true, nil
+			}); err != nil {
+				t.Fatal("got unexpected result")
```
Comment (pohly) [inline pkg/scheduler/scheduler_test.go:1071]: Hrrrm. `wait.Poll` is not a good way to poll in tests (see "Good code that waits for something to happen meets the following criteria" in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts).

If this times out, then all that shows up in the unit test output is "got unexpected result". At least include the error, which should be the one from `fmt.Errorf`.

Comment (pohly) [reply pkg/scheduler/scheduler_test.go:1071]: You could switch to k/k/test/utils/ktesting and use https://pkg.go.dev/k8s.io/kubernetes/test/utils/ktesting#Eventually. But that's a bigger change.

Up to you...

Comment (pohly) [reply pkg/scheduler/scheduler_test.go:1071]: Actually... does `wait.Poll` really poll here? As far as I can tell, your callback is getting invoked *once* and either returns "true" or an error. In both cases [wait.PollUntilContextCancel](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextCancel) returns immediately, without retrying later. So this is the same as before, just with a less informative failure?

Comment (liggitt) [reply pkg/scheduler/scheduler_test.go:1071]: agreed, this seems racy both before and after this change

File: pkg/scheduler/scheduler_test.go (hunk 1 above)
Comment (liggitt) [reply pkg/scheduler/scheduler_test.go:1007]: what is the longest we would expect this to take to settle to the expected result? I would hope a few seconds, not a minute and a half

Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go:1007]: Generally permit plugin will wait for a relative long time like 100s but looking into the codes, only used in one case and no waiting logic added, 5s seems enough, I tested locally, under race mode, almost 3s. But to be more tolerant, so 1) I set this to 10s and 2) also make it a constant.

File: pkg/scheduler/scheduler_test.go (hunk 2 above)
Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go]: It's a mistake, fixed.

Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go]: PTAL

File: pkg/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -994,15 +997,20 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
...
 				fakePermit: newFakePermitPlugin(eventRecorder),
 			}
 
 			_, ctx := ktesting.NewTestContext(t)
-			ctx, cancel := context.WithCancel(ctx)
+			// Theoretically, the timeout should be greater than permitTimeout, but considering
+			// we have no waiting logic in permit plugin, this is fair enough.
+			ctx, cancel := context.WithTimeout(ctx, permitTimeout)
 			defer cancel()
+			tCtx := utiltesting.InitCtx(ctx, t)
```
Comment (pohly) [inline pkg/scheduler/scheduler_test.go:1013]: You don't need both. You can remove this:
```
_, ctx := ktesting.NewTestContext(t)
...
ctx, cancel := context.WithTimeout(ctx, permitTimeout)
```

Then below:
```
scheduler, err := New(tCtx, ...
```

This removes the `permitTimeout`. I don't understand why you are adding it to the context of the scheduler.

The test is going to be stuck below in the poll loop until it succeeds or times out there, Giving the scheduler a context with a separate timeout does not change that.

Or was this timeout intended for the polling? Then below use:
```
utiltesting.Eventually(tCtx, ...).Timeout(permitTimeout).Should(...)
```

Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go]: Eventually timeout is all I need.

Comment (kerthcet) [reply pkg/scheduler/scheduler_test.go]: Fixed.

**Review by pohly — COMMENTED** [review]
/lgtm

**Pull request merged**